### PR TITLE
Fixed sampling

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:
Added fixed sampling option to `StreamingDataset` which makes sure that the samples are the same across epochs.

Enables users to keep samples consistent across epochs by keeping the random seed the same between epochs.

## Issue #, if available:
STR-36

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -284,12 +292,25 @@ def __init__(self,
...
         self.shuffle_block_size = shuffle_block_size
+        self.sampling_method = sampling_method.lower().strip()
 
         # Check streams vs remote/local.
         if bool(streams) == (bool(remote) or bool(local)):
             raise ValueError(
                 'You must provide either `streams` or `remote`/`local`, but not both.')
 
+        # Check sampling method is one of "balanced" or "fixed".
+        if sampling_method not in ['balanced', 'fixed']:
```
Comment (karan6181) [inline streaming/base/dataset.py:303]: nit: `sampling_method` -> `self.sampling_method` !!

File: tests/test_streaming.py [hunk 2]
```diff
@@ -12,6 +12,103 @@
...
+        assert samples_seen == epoch_size - (epoch_size % batch_size)
+    else:
+        assert samples_seen == epoch_size
+
+
+@pytest.mark.parametrize('batch_size', [1, 4])
+@pytest.mark.parametrize('seed', [2222])
+@pytest.mark.parametrize('shuffle', [False, True])
+@pytest.mark.parametrize('drop_last', [False])
+@pytest.mark.parametrize('num_canonical_nodes', [1])
```
Comment (karan6181) [inline tests/test_streaming.py:60]: Can we have a variation in `num_canonical_nodes` ? such as `4` or `8` ? Plus increase the `num_samples` if it is necessary!

File: tests/test_streaming.py [hunk 3]
```diff
@@ -12,6 +12,103 @@
...
+                               batch_size=batch_size,
+                               shuffle_seed=seed,
+                               num_canonical_nodes=num_canonical_nodes,
+                               epoch_size=epoch_size,
+                               sampling_method='fixed')
+
+    # Build DataLoader
+    dataloader = StreamingDataLoader(dataset=dataset,
+                                     batch_size=batch_size,
+                                     num_workers=0,
```
Comment (karan6181) [inline tests/test_streaming.py:85]: Any reason for setting this `num_workers=0` ? Would recommend setting this to real world value such as `4` or `8` !!

File: tests/test_streaming.py [hunk 4]
```diff
@@ -12,6 +12,103 @@
...
+                                     drop_last=drop_last)
+
+    # iterate once over the dataloader (first epoch)
```
Comment (karan6181) [inline tests/test_streaming.py:88]: You can simplify the two for loop into one something like this: I haven't ran it, maybe you need to take care of shallow copy / deep copy !!

```
# iterate once over the dataloader (first epoch)
first_samples_seen = {}
second_samples_seen = {}
is_first_epoch = True
for batch in dataloader:
    samples_seen = first_samples_seen if is_first_epoch else second_samples_seen
    for element in batch['sample']:
        int_element = int(element)
        if int_element in samples_seen:
            samples_seen[int_element] += 1
        else:
            samples_seen[int_element] = 1
    samples_seen = {}
```

Comment (snarayan21) [reply tests/test_streaming.py:88]: The dataloader only has enough batches for a single epoch I thought? so you would have to loop twice right? To my understanding the dataloader restarts at the beginning of the epoch when iterated over again. I'll definitely remove the redundancy in the for loops though. The elements are the just ints and dictionary equality should work...
